agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,481 @@
+	if err := os.RemoveAll(agentTmp); err != nil {
+		return nil, fmt.Errorf("removing agent tmp from previous runs: %w", err)
+	}
+
+	if err := os.MkdirAll(agentTmp, 0o700); err != nil {
+		return nil, fmt.Errorf("preparing agent tmp: %w", err)
+	}
+
+	return &PodmanClient{
+		log:        logrus.WithField("component", "podman"),
+		opts:       opts,
+		wg:         waitgroupx.WithContext(context.Background()),
+		containers: make(map[cproto.ID]*PodmanContainer),
+	}, nil
+}
+
+// Close the client, killing all running containers and removing our scratch space.
+func (s *PodmanClient) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Since we launch procs with exec.CommandContext under s.wg's context, this cleans them up.
+	s.wg.Close()
+
+	if err := os.RemoveAll(agentTmp); err != nil {
+		return fmt.Errorf("cleaning up agent tmp: %w", err)
+	}
+	return nil
+}
+
+func getPullCommand(req docker.PullImage, image string) (string, []string) {
+	// C.f. singularity where if req.ForcePull is set then a 'pull --force' is done.
+	// podman does not have this option, though it does have '--pull always' on the
+	// run command.
+	return "podman", []string{"pull", image}
+}
+
+// PullImage implements container.ContainerRuntime.
+func (s *PodmanClient) PullImage(
+	ctx context.Context,
+	req docker.PullImage,
+	p events.Publisher[docker.Event],
+) (err error) {
+	return cruntimes.PullImage(ctx, req, p, &s.wg, s.log, getPullCommand)
+}
+
+// CreateContainer implements container.ContainerRuntime.
+func (s *PodmanClient) CreateContainer(
+	ctx context.Context,
+	id cproto.ID,
+	req cproto.RunSpec,
+	p events.Publisher[docker.Event],
+) (string, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.containers[id] = &PodmanContainer{Req: req}
+	return id.String(), nil
+}
+
+// RunContainer implements container.ContainerRuntime.
+// nolint: golint,maintidx // Both contexts can't both be first / TODO refactor.
+func (s *PodmanClient) RunContainer(
+	ctx context.Context,
+	waitCtx context.Context,
+	id string,
+	p events.Publisher[docker.Event],
+) (*docker.Container, error) {
+	s.mu.Lock()
+	var cont *PodmanContainer
+	for cID, rcont := range s.containers {
+		if cproto.ID(id) != cID {
+			continue
+		}
+		cont = rcont
+		break
+	}
+	s.mu.Unlock()
+
+	if cont == nil {
+		return nil, container.ErrMissing
+	}
+	req := cont.Req
+
+	u, err := user.Current()
+	if err != nil {
+		return nil, fmt.Errorf("checking user: %w", err)
+	}
+
+	uidgid := fmt.Sprintf("%s:%s", u.Uid, u.Gid)
+	if req.ContainerConfig.User != uidgid {
+		return nil, fmt.Errorf(
+			"agent running as %s cannot launch as user %s",
+			uidgid, req.ContainerConfig.User,
+		)
+	}
+
+	tmpdir, err := os.MkdirTemp(agentTmp, fmt.Sprintf("*-%s", id))
+	if err != nil {
+		return nil, fmt.Errorf("making tmp dir for archives: %w", err)
+	}
+
+	var args []string
+	args = append(args, "run")
+	args = append(args, "--rm")
+	// args = append(args, "--writable-tmpfs")
```

Comment:
Why is this commented out?

Reply:
Left over from singularity -- I'll remove it.